### PR TITLE
Sprint 2. TC1349 - User can mark cards for re-checking TC1350-01 User can mark the card TC1350-02 After marking the card, the number of marked cards increased by 1.

### PR DESCRIPTION
**Pull request by LaysanNova**
Sprint 2. TC1349 - User can mark cards for re-checking
!class TestTutorPage > method clickAddToFlashCard
TC1350-01 User can mark the card
TC1350-02 After marking the card, the number of marked cards increased by 1.

### CODE REVIEW

File: src/test/java/pages/BaseLocator.java [hunk 1]
```diff
@@ -56,7 +56,15 @@ protected Locator image(String text) {
...
 
+    protected Locator textBox() {
+        return getPage().getByRole(AriaRole.TEXTBOX);
+    }
+
     protected void clickButton(String text) {
         button(text).click();
     }
+
+    protected void clickSubject(String subject) {
```
Comment (ArinaJur) [inline src/test/java/pages/BaseLocator.java:67]: This is not a generic method. Please put it into the POM class where you use it (FlashcardsPage, etc)

File: src/test/java/pages/FlashcardsPage.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
+package pages;
+
+import com.microsoft.playwright.Locator;
+import com.microsoft.playwright.Page;
+import com.microsoft.playwright.Playwright;
+import com.microsoft.playwright.options.AriaRole;
+
+public class FlashcardsPage extends BaseLocator{
+    public Locator markedForRechecking = getPage().getByRole(AriaRole.BUTTON)
```
Comment (ArinaJur) [inline src/test/java/pages/FlashcardsPage.java:9]: All locators should remain private final.

File: src/test/java/pages/FlashcardsPage.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+            .filter(new Locator.FilterOptions().setHas(getPage().getByText("Marked for re-checking")));
```
Comment (ArinaJur) [inline src/test/java/pages/FlashcardsPage.java:10]: Please try to refactor this lines to something like "button().filter("Marked for re-checking")"

File: src/test/java/pages/FlashcardsPage.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+    public Locator numberMarkedForRechecking = getPage().getByRole(
```
Comment (ArinaJur) [inline src/test/java/pages/FlashcardsPage.java:12]: Is it possible to use "button("Marked for re-checking").locator("//div[2]");" ?

Also please try to refactor "//div[2]" to CSS selector.

File: src/test/java/pages/FlashcardsPage.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+            AriaRole.BUTTON, new Page.GetByRoleOptions().setName("Marked for re-checking"))
+            .locator("//div[2]");
+    public Locator flashcardsMenuButton = button("Flashcards");
```
Comment (ArinaJur) [inline src/test/java/pages/FlashcardsPage.java:15]: Let's refer to this as flashcardsMenu

File: src/test/java/pages/FlashcardsPage.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+    public FlashcardsPage(Page page, Playwright playwright) {
+        super(page, playwright);
+    }
+
+
+    public FlashcardsPage clickFlashcardsMenuButton() {
```
Comment (ArinaJur) [inline src/test/java/pages/FlashcardsPage.java:22]: clickFlashcardsMenu()

File: src/test/java/pages/TestTutorPage.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+public class TestTutorPage extends BaseLocator {
+
+    public Locator addToFlashcardButton = button("Add to flashcard");
+    public Locator removeFromFlashcards = button("Remove from flashcards");
+    public Locator markForReview = button("Mark for review");
+    public Locator removeFromMarked = button("Remove from marked");
+    public Locator endButton = exactButton("End");
+    public Locator yesButton = exactButton("Yes");
+    public Locator skipButton = exactButton("Skip");
+    public Locator closeTestButton = exactButton("Close the test");
```
Comment (ArinaJur) [inline src/test/java/pages/TestTutorPage.java:16]: Please make all locators private final

File: src/test/java/pages/TestTutorPage.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
+package pages;
+
+import com.microsoft.playwright.Locator;
+import com.microsoft.playwright.Page;
+import com.microsoft.playwright.Playwright;
+
+public class TestTutorPage extends BaseLocator {
+
+    public Locator addToFlashcardButton = button("Add to flashcard");
+    public Locator removeFromFlashcards = button("Remove from flashcards");
```
Comment (ArinaJur) [inline src/test/java/pages/TestTutorPage.java:10]: rremoveFromFlashcardsButton

File: src/test/java/pages/TestTutorPage.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    public Locator markForReview = button("Mark for review");
```
Comment (ArinaJur) [inline src/test/java/pages/TestTutorPage.java:11]: markForReviewButton

File: src/test/java/pages/TestTutorPage.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    public Locator removeFromMarked = button("Remove from marked");
```
Comment (ArinaJur) [inline src/test/java/pages/TestTutorPage.java:12]: removeFromMarkedButton

File: src/test/java/pages/TestTutorPage.java (hunk 7 above)
Comment (ArinaJur) [inline src/test/java/pages/TestTutorPage.java:16]: closeTheTestButton

File: src/test/java/pages/TestTutorPage.java [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+    public Locator endButton = exactButton("End");
+    public Locator yesButton = exactButton("Yes");
+    public Locator skipButton = exactButton("Skip");
+    public Locator closeTestButton = exactButton("Close the test");
+
+    public TestTutorPage(Page page, Playwright playwright) {
+        super(page, playwright);
+    }
+
+    public TestTutorPage clickAddToFlashCard() {
```
Comment (ArinaJur) [inline src/test/java/pages/TestTutorPage.java:22]: clickAddToFlashCardButton()

File: src/test/java/pages/TestTutorPage.java [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+        yesButton.click();
+        return this;
+    }
+
+    public TestTutorPage clickSkipButton() {
+        skipButton.click();
+        return this;
+    }
+
+    public void clickCloseTestButton() {
```
Comment (ArinaJur) [inline src/test/java/pages/TestTutorPage.java:45]: clickCloseTheTestButton()

File: src/test/java/pages/TestsPage.java [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+
+import java.util.List;
+import static utils.TestUtils.getRandomValue;
+
+public class TestsPage extends BaseLocator {
+    public Locator testsMenuButton = button("Tests");
+    public Locator numberOfQuestions = textBox();
+    public Locator generateStartButton = button("Generate & Start");
+    public Locator cancelButtonOnPromptWindow = button("Cancel");
+    public Locator promptMessageUnfinishedTest  = text("You have unfinished test. Do you want to continue?");
```
Comment (ArinaJur) [inline src/test/java/pages/TestsPage.java:15]: Please make all locators private final

File: src/test/java/pages/TestsPage.java [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+
+import com.microsoft.playwright.Locator;
+import com.microsoft.playwright.Page;
+import com.microsoft.playwright.Playwright;
+
+import java.util.List;
+import static utils.TestUtils.getRandomValue;
+
+public class TestsPage extends BaseLocator {
+    public Locator testsMenuButton = button("Tests");
```
Comment (ArinaJur) [inline src/test/java/pages/TestsPage.java:11]: testsMenu

File: src/test/java/pages/TestsPage.java [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+    public Locator numberOfQuestions = textBox();
```
Comment (ArinaJur) [inline src/test/java/pages/TestsPage.java:12]: numberOfQuestionsTextBox

File: src/test/java/pages/TestsPage.java [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+    public Locator generateStartButton = button("Generate & Start");
```
Comment (ArinaJur) [inline src/test/java/pages/TestsPage.java:13]: generateAndStartButton

File: src/test/java/pages/TestsPage.java [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+    public Locator cancelButtonOnPromptWindow = button("Cancel");
```
Comment (ArinaJur) [inline src/test/java/pages/TestsPage.java:14]: cancelButton

File: src/test/java/pages/TestsPage.java (hunk 13 above)
Comment (ArinaJur) [inline src/test/java/pages/TestsPage.java:15]: unfinishedTestText

File: src/test/java/pages/TestsPage.java [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+    public Locator promptMessageUnfinishedTest  = text("You have unfinished test. Do you want to continue?");
+
+    public TestsPage(Page page, Playwright playwright) {
+        super(page, playwright);
+    }
+
+    protected String catchSubject() {
+
+        List<Locator> subjects = getPage().locator("div.sc-gHnmx.iEQmLh").all();
```
Comment (ArinaJur) [inline src/test/java/pages/TestsPage.java:23]: Please move this to locators fields

File: src/test/java/pages/TestsPage.java [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+
+        String text = getRandomValue(subjects).textContent();
```
Comment (ArinaJur) [inline src/test/java/pages/TestsPage.java:25]: It is better to use TestUtils.getRandomValue()

File: src/test/java/pages/TestsPage.java [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+    public Locator numberOfQuestions = textBox();
+    public Locator generateStartButton = button("Generate & Start");
+    public Locator cancelButtonOnPromptWindow = button("Cancel");
+    public Locator promptMessageUnfinishedTest  = text("You have unfinished test. Do you want to continue?");
+
+    public TestsPage(Page page, Playwright playwright) {
+        super(page, playwright);
+    }
+
+    protected String catchSubject() {
```
Comment (ArinaJur) [inline src/test/java/pages/TestsPage.java:21]: getSubject()

File: src/test/java/pages/TestsPage.java [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+
+        String text = getRandomValue(subjects).textContent();
+
+        String substring = text.substring(text.lastIndexOf('('));
+
+        return text.replace(substring, "").trim();
+
+    }
+
+    public void initiateTest() {
```
Comment (ArinaJur) [inline src/test/java/pages/TestsPage.java:33]: This method should be refactored to use smaller test scenario steps

clickTestsMenu()
dismissPromptIfShown()
clickSubject(getSubject())
inputNumberOfQuestions()
clickGenerateAndStartButton()

etc

File: src/test/java/tests/TestsTest.java [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+
+import pages.TestTutorPage;
+import pages.TestsPage;
+
+import static com.microsoft.playwright.assertions.PlaywrightAssertions.assertThat;
+import static java.lang.Integer.parseInt;
+
+public class TestsTest extends BaseTest {
+
+    public String preconditions_1349() {
```
Comment (ArinaJur) [inline src/test/java/tests/TestsTest.java:16]: This method should be moved to the POM and given a name that clearly describes its purpose (for ex getNumberOfCardsBeforeTestStarted)

File: src/test/java/tests/TestsTest.java [hunk 23]
```diff
@@ -0,0 +1,39 @@
...
+        String numberOfCardsForReChecking = new FlashcardsPage(getPage(), getPlaywright())
+                .getNumberOfCardsForReChecking();
+        new TestsPage(getPage(), getPlaywright()).initiateTest();
+
+        return numberOfCardsForReChecking;
+    }
+    @Test
+    public void test_1349() {
```
Comment (ArinaJur) [inline src/test/java/tests/TestsTest.java:24]: Please name the tests according to the corresponding test case name.

File: src/test/java/tests/TestsTest.java [hunk 24]
```diff
@@ -0,0 +1,39 @@
...
+        String numberOfCardsForReChecking = preconditions_1349();
+
+        new TestTutorPage(getPage(), getPlaywright())
+                .clickAddToFlashCard()
+                .clickEndButton()
+                .clickYesButton()
+                .clickSkipButton()
+                .clickCloseTestButton();
+
+        Locator newNumberOfCardsForReChecking = new FlashcardsPage(getPage(), getPlaywright()).clickFlashcardsMenuButton().numberMarkedForRechecking;
```
Comment (ArinaJur) [inline src/test/java/tests/TestsTest.java:35]: After refactoring the code above, we should reconsider the test format.

File: src/test/java/tests/TestsTest.java [hunk 25]
```diff
@@ -0,0 +1,39 @@
...
+
+        assertThat(newNumberOfCardsForReChecking).hasText(Integer.toString(parseInt(numberOfCardsForReChecking) + 1));
```
Comment (ArinaJur) [inline src/test/java/tests/TestsTest.java:37]: Please always remove actions or calculations from assert

File: src/test/java/utils/TestUtils.java [hunk 26]
```diff
@@ -2,9 +2,16 @@
 
 import com.microsoft.playwright.Locator;
 
+import java.util.List;
 import java.util.Random;
 
 public class TestUtils {
+    public static  <T> T getRandomValue(List<T> listValues) {
```
Comment (ArinaJur) [inline src/test/java/utils/TestUtils.java:9]: It should be a separate PR

**Review by ArinaJur — CHANGES_REQUESTED** [review]

File: src/test/java/pages/TestsPage.java (hunk 13 above)
Comment (LaysanNova) [reply src/test/java/pages/TestsPage.java:15]: не совсем поняла

**Comment by LaysanNova** [comment]
не актуально

**Comment by LaysanNova** [comment]
не актуально

**Pull request closed without merging**